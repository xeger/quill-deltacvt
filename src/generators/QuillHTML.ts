import BaseHTML, { EmbedFormatter, TextFormatter } from './BaseHTML';

/**
 * HTML generators for non-textual content (e.g. images)
 * in a chunk.
 */
const EMBEDS: Record<string, EmbedFormatter> = {
  image(src, attributes) {
    const { width, height } = attributes;
    const optWidth = width ? ` width="${width}"` : '';
    const optHeight = height ? ` height="${height}"` : '';

    return `<img src="${src}"${optWidth}${optHeight}/>`;
  },
};

/**
 * HTML generators for character formats. Line formats are context-sensitive
 * and need to be handled as part of the generation algorithm.
 */
const TEXTS: Record<string, TextFormatter> = {
  bold: (span) => {
    span.styles.push('font-weight:bold');
  },
  color: (span, color) => {
    span.styles.push(`color:${color}`);
  },
  font: (span, font) => {
    span.styles.push(`font-family:'${font}'`);
  },
  italic: (span) => {
    span.styles.push('font-style:italic');
  },
  link: (span, href) => {
    span.text = `<a href="${href}">${span.text}</a>`;
  },
  size: (span, size) => {
    span.styles.push(`font-size:${size}`);
  },
  underline: (span) => {
    span.styles.push('text-decoration:underline');
  },
};

export interface Options {
  strict?: true;
}

interface Line {
  align?: string;
  list?: string;
  text: string;
}

/**
 * Generator that outputs HTML approximating the exact HTML output
 * of Quill's Parchment formats. No attempt is made to match the
 * visual appearance of the Quill editor.
 */
export default class QuillHTML extends BaseHTML {
  embedFormatters: Record<string, EmbedFormatter> = { ...EMBEDS };
  strict?: true;
  textFormatters: Record<string, TextFormatter> = { ...TEXTS };

  constructor(options: Options = {}) {
    super(EMBEDS, TEXTS, options);
  }

  formatLine(current: Line, prior: Line): string {
    const { align, list, text } = current;
    const style = align ? ` style="text-align: ${align}"` : '';

    let html: string;
    if (list) html = `<li${style}>${text}</li>`;
    else if (text) {
      if (text === '\n') html = '<br/>';
      else html = `<p${style}>${text}</p>`;
    } else html = '';

    if (list !== prior?.list) {
      const prefix: string[] = [];
      if (prior?.list === 'bullet') prefix.push(`</ul>`);
      else if (prior?.list === 'ordered') prefix.push(`</ol>`);
      if (list === 'bullet') prefix.push(`<ul>`);
      else if (list === 'ordered') prefix.push(`<ol>`);
      html = `${prefix.join('')}${html}`;
    }

    return html;
  }

  isLineFormat(format: string): boolean {
    switch (format) {
      case 'align':
      case 'list':
        return true;
      default:
        return false;
    }
  }

  wrap(fragment: string): string {
    return `<!DOCTYPE html><html><body>${fragment}</body></html>`;
  }
}
