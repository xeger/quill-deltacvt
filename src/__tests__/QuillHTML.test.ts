import 'html-validate/jest';

import * as visual from './visual';
import { generateFragment, generate } from '..';
import QuillHTML from '../generators/QuillHTML';
import {
  FONT,
  IMAGES,
  LISTS,
  LISTS_STYLED,
  MALICIOUS,
  MULTI_NEWLINE,
  SMORGASBORD,
  TRIVIAL_ALIGN,
  TRIVIAL_LIST,
} from '../../test/fixtures';

const UNKNOWN_ATTRIBUTE = [{ attributes: { banana: true }, insert: 'hi' }];
const UNKNOWN_EMBED = [{ insert: { banana: 'banana' } }];
const UNKNOWN_INSERT = [{ insert: 42 as any }];

// Generate and validate HTML from ops.
function genHTML(ops) {
  const html = generate(ops, new QuillHTML());
  expect(html).toHTMLValidate();
  return html;
}

describe('generators/QuillHTML', () => {
  it('handles unknown content', () => {
    const g = new QuillHTML();
    expect(generateFragment(UNKNOWN_ATTRIBUTE, g)).toEqual('<p>hi</p>');
    expect(generateFragment(UNKNOWN_EMBED, g)).toEqual('');
    expect(generateFragment(UNKNOWN_INSERT, g)).toEqual('');
  });

  describe('algorithm', () => {
    it('handles images', () => {
      visual.matchSnapshot(genHTML(IMAGES));
    });

    it('escapes HTML syntax', () => {
      visual.matchSnapshot(genHTML(MALICIOUS));
    });

    it('handles fonts', () => {
      visual.matchSnapshot(genHTML(FONT));
    });

    it('handles lists', () => {
      visual.matchSnapshot(genHTML(LISTS));
    });

    it('handles styled lists', () => {
      visual.matchSnapshot(genHTML(LISTS_STYLED));
    });

    it('handles trivial inputs', () => {
      visual.matchSnapshot(genHTML(TRIVIAL_LIST));
    });

    it('handles standard Quill formats', () => {
      visual.matchSnapshot(genHTML(SMORGASBORD));
    });

    it('handles trivial aligned text', () => {
      visual.matchSnapshot(genHTML(TRIVIAL_ALIGN));
    });

    it('handles line formats within the same DeltaOperation', () => {
      visual.matchSnapshot(genHTML(MULTI_NEWLINE));
    });
  });

  describe('options', () => {
    describe('strict mode', () => {
      const g = new QuillHTML({ strict: true });

      it('gives helpful embed messages', () => {
        expect(() => generateFragment(UNKNOWN_EMBED, g)).toThrowError(
          'Unknown embed: banana'
        );
      });

      it('gives helpful attribute messages', () => {
        expect(() => generateFragment(UNKNOWN_ATTRIBUTE, g)).toThrowError(
          'Unknown attribute: banana'
        );
      });

      it('gives helpful insert messages', () => {
        expect(() => generateFragment(UNKNOWN_INSERT, g)).toThrowError(
          'Unknown insert type: number'
        );
      });
    });
  });
});
