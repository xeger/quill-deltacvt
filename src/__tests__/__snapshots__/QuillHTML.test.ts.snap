// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generators/QuillHTML algorithm escapes HTML syntax 1`] = `
"<!DOCTYPE html><html><body><p>&lt;script&gt;alert(&quot;xss&quot;);&lt;/script&gt;
</p><ul><li>&lt;script&gt;alert(&quot;xss&quot;);&lt;/script&gt;
</li><li style=\\"text-align: right\\">&lt;script&gt;alert(&quot;xss&quot;);&lt;/script&gt;
</li><li>not xss &amp; okay
</li><li style=\\"text-align: center\\">not xss &amp; okay
</li></ul><br/></body></html>"
`;

exports[`generators/QuillHTML algorithm handles images 1`] = `
"<!DOCTYPE html><html><body><p><img src=\\"https://quilljs.com/assets/images/logo.svg\\" width=\\"140\\" height=\\"38\\"/>
</p><p>Sized image.
</p><br/><p><img src=\\"https://raw.githubusercontent.com/quilljs/quill/master/assets/favicon.png\\"/>
</p><p>Left-aligned image.
</p><br/><p style=\\"text-align: center\\"><img src=\\"https://raw.githubusercontent.com/quilljs/quill/master/assets/favicon.png\\"/>
</p><p style=\\"text-align: center\\">Center-aligned image.
</p><br/><p style=\\"text-align: right\\"><img src=\\"https://raw.githubusercontent.com/quilljs/quill/master/assets/favicon.png\\"/>
</p><p style=\\"text-align: right\\">Right-aligned image.
</p><br/><p style=\\"text-align: center\\"><img src=\\"https://quilljs.com/assets/images/logo.svg\\" width=\\"140\\" height=\\"38\\"/>
</p><p style=\\"text-align: center\\">Sized and centered image.
</p></body></html>"
`;

exports[`generators/QuillHTML algorithm handles line formats within the same DeltaOperation 1`] = `
"<!DOCTYPE html><html><body><p style=\\"text-align: center\\">Centered.
</p><p style=\\"text-align: center\\">Centered <span style=\\"font-weight:bold;color:#b2b200\\">with</span> <span style=\\"text-decoration:underline\\"><a href=\\"http://quilljs.com\\">some</a></span> formatting.
</p><p style=\\"text-align: right\\">Right aligned.
</p><p style=\\"text-align: right\\"><span style=\\"color:#66b966\\">Right</span> <span style=\\"font-style:italic;text-decoration:underline\\">aligned</span> with <span style=\\"font-weight:bold\\">some</span> formatting.
</p><br/><p style=\\"text-align: justify\\"><span style=\\"font-weight:bold\\">Justified</span> so the text nicely fills the line. <span style=\\"color:#f06666\\">Justified</span> <span style=\\"color:#66a3e0\\">so</span> <span style=\\"color:#c285ff\\">the</span> text nicely fills the line. Justified so the text nicely fills the line.
</p><p style=\\"text-align: justify\\">Also justified.
</p><br/><ul><li>Unordered
</li><li>List
</li></ul><br/><ol><li>Ordered
</li><li>List
</li></ol><br/><ul><li>Mixed
</li></ul><ol><li>List
</li></ol><ul><li>Weirdly
</li></ul><ol><li>Laid Out
</li><li>With Numbers
</li></ol><ul><li>Interspersed
</li></ul><br/><ul><li>list
</li><li style=\\"text-align: center\\">with
</li><li style=\\"text-align: right\\">alignment
</li><li style=\\"text-align: justify\\">and justification in case the list items get stupendously enormously long
</li></ul></body></html>"
`;

exports[`generators/QuillHTML algorithm handles lists 1`] = `
"<!DOCTYPE html><html><body><p>Normal text to start.
</p><ul><li>chalk
</li><li>
</li><li>cheese
</li></ul><p>Now an ordered list:
</p><ol><li>apples
</li><li style=\\"text-align: right\\">oranges
</li><li>pears
</li><li>
</li></ol><p>And finally, adjacent lists:
</p><ul><li>please
</li><li>excuse
</li><li>my
</li></ul><ol><li>dear
</li><li>aunt
</li><li>sally
</li></ol></body></html>"
`;

exports[`generators/QuillHTML algorithm handles standard Quill formats 1`] = `
"<!DOCTYPE html><html><body><p>Plain.  <span style=\\"font-weight:bold\\">Bold. </span><span style=\\"font-style:italic\\">Italic. </span><span style=\\"text-decoration:underline\\">Underline.</span> <span style=\\"color:#9933ff\\">Colored.</span> <a href=\\"https://quilljs.com\\">Link.</a>
</p><p><span style=\\"font-size:x-small\\">Small, </span>normal, <span style=\\"font-size:x-large\\">large, </span><span style=\\"font-size:xx-large\\">huge!</span>
</p><p>Normal family. <span style=\\"font-family:'monospace'\\">Monospace family.</span> <span style=\\"font-family:'serif'\\">Serif family.</span>
</p><p>Left aligned.
</p><p>Left <span style=\\"font-weight:bold\\">aligned</span> with <span style=\\"font-style:italic;text-decoration:underline\\">some</span> <span style=\\"color:#e60000\\">formatting</span>.
</p><p style=\\"text-align: center\\">Centered.
</p><p style=\\"text-align: center\\">Centered <span style=\\"font-weight:bold;color:#b2b200\\">with</span> <span style=\\"text-decoration:underline\\"><a href=\\"http://quilljs.com\\">some</a></span> formatting.
</p><p style=\\"text-align: right\\">Right aligned.
</p><p style=\\"text-align: right\\"><span style=\\"color:#66b966\\">Right</span> <span style=\\"font-style:italic;text-decoration:underline\\">aligned</span> with <span style=\\"font-weight:bold\\">some</span> formatting.
</p><br/><p style=\\"text-align: justify\\"><span style=\\"font-weight:bold\\">Justified</span> so the text nicely fills the line. <span style=\\"color:#f06666\\">Justified</span> <span style=\\"color:#66a3e0\\">so</span> <span style=\\"color:#c285ff\\">the</span> text nicely fills the line. Justified so the text nicely fills the line.
</p><p style=\\"text-align: justify\\">Also justified.
</p><br/><ul><li>Unordered
</li><li>List
</li></ul><br/><ol><li>Ordered
</li><li>List
</li></ol><br/><ul><li>Mixed
</li></ul><ol><li>List
</li></ol><ul><li>Weirdly
</li></ul><ol><li>Laid Out
</li><li>With Numbers
</li></ol><ul><li>Interspersed
</li></ul><br/><ul><li>list
</li><li style=\\"text-align: center\\">with
</li><li style=\\"text-align: right\\">alignment
</li><li style=\\"text-align: justify\\">and justification in case the list items get stupendously enormously long
</li></ul><br/></body></html>"
`;

exports[`generators/QuillHTML algorithm handles styled lists 1`] = `
"<!DOCTYPE html><html><body><ul><li><span style=\\"font-weight:bold;text-decoration:underline\\">lots</span> <span style=\\"color:#9933ff\\">of</span> <span style=\\"font-weight:bold;font-style:italic;text-decoration:underline\\"><a href=\\"http://google.com\\">style</a></span> <span style=\\"font-style:italic\\">happening</span>
</li><li>lots <span style=\\"font-weight:bold\\">of</span> <span style=\\"color:#e60000\\">function</span> too
</li><li>gotta love it all 
</li></ul></body></html>"
`;

exports[`generators/QuillHTML algorithm handles trivial aligned text 1`] = `
"<!DOCTYPE html><html><body><p style=\\"text-align: center\\">hi
</p></body></html>"
`;

exports[`generators/QuillHTML algorithm handles trivial inputs 1`] = `
"<!DOCTYPE html><html><body><ul><li>hi
</li></ul></body></html>"
`;
